js/RTCPeerConnection.ts

```diff
@@ -1,878 +1,913 @@
+		void provider;
+		void options;
+		throw new Error('RTCPeerConnection.setIdentityProvider not implemented');
+	}
+	generateCertificate(keygenAlgorithm: AlgorithmIdentifier): Promise<RTCCertificate> {
+		void keygenAlgorithm;
+		throw new Error('RTCPeerConnection.generateCertificate not implemented');
+	}
+	getDefaultIceServers(): RTCIceServer[] {
+		throw new Error('RTCPeerConnection.getDefaultIceServers not implemented');
+	}
+}
 
 // Save current RTCPeerConnection.prototype
-RTCPeerConnection.prototype_descriptor = Object.getOwnPropertyDescriptors(
-	RTCPeerConnection.prototype
+(RTCPeerConnectionShim as any).prototype_descriptor = Object.getOwnPropertyDescriptors(
+	RTCPeerConnectionShim.prototype
 );
 
-/**
- * Private API.
- */
-function fixPcConfig(pcConfig) {
+function fixPcConfig(pcConfig: RTCConfiguration) {
 	if (!pcConfig) {
 		return {
 			iceServers: []
 		};
 	}
 
-	var iceServers = pcConfig.iceServers,
-		i,
-		len,
-		iceServer;
+	const iceServers = pcConfig.iceServers;
 
 	if (!Array.isArray(iceServers)) {
 		pcConfig.iceServers = [];
 		return pcConfig;
 	}
 
-	for (i = 0, len = iceServers.length; i < len; i++) {
-		iceServer = iceServers[i];
-
+	iceServers.forEach((iceServer: RTCIceServer & { url?: string }) => {
 		// THe Objective-C wrapper of WebRTC is old and does not implement .urls.
 		if (iceServer.url) {
-			continue;
+			return;
 		} else if (Array.isArray(iceServer.urls)) {
 			iceServer.url = iceServer.urls[0];
 		} else if (typeof iceServer.urls === 'string') {
 			iceServer.url = iceServer.urls;
 		}
-	}
+	});
 
 	return pcConfig;
 }
-
-function isClosed() {
-	return this.signalingState === 'closed';
-}
-
-function onEvent(data) {
-	var type = data.type,
-		self = this,
-		event = new Event(type),
-		dataChannel,
-		id;
-
-	Object.defineProperty(event, 'target', { value: self, enumerable: true });
-
-	debug('onEvent() | [type:%s, data:%o]', type, data);
-
-	switch (type) {
-		case 'signalingstatechange':
-			this.signalingState = data.signalingState;
-			break;
-
-		case 'icegatheringstatechange':
-			this.iceGatheringState = data.iceGatheringState;
-			break;
-
-		case 'iceconnectionstatechange':
-			this.iceConnectionState = data.iceConnectionState;
-
-			// Emit "connected" on remote streams if ICE connected.
-			if (data.iceConnectionState === 'connected') {
-				for (id in this.remoteStreams) {
-					if (this.remoteStreams.hasOwnProperty(id)) {
-						this.remoteStreams[id].emitConnected();
-					}
-				}
-			}
-			break;
-
-		case 'icecandidate':
-			if (data.candidate) {
-				event.candidate = new RTCIceCandidate(data.candidate);
-			} else {
-				event.candidate = null;
-			}
-			// Update _localDescription.
-			if (this._localDescription) {
-				this._localDescription.type = data.localDescription.type;
-				this._localDescription.sdp = data.localDescription.sdp;
-			} else {
-				this._localDescription = new RTCSessionDescription(data);
```

Comment:
Another example of a bug caught by TypeScript.  I think `this._localDescription` is usually set, but if it wasn't, this would have created an invalid session description.  Should have used `data.localDescription` instead.

Reply:
Not sure about that one.
Never had issue with this before.

Reply:
`RTCSessionDescription` expects something like `{ type: string, sdp: string }`.  If you look at the lines right above it, `data.localDescription` matches that type, but `data` doesn't.  Like I said, it may be rare that `this._localDescription` is not set, thus the bug wouldn't show up.

Reply:
its a Type thing, but data contains type/sdp whe used line 821.
For this 2 "Another example of a bug caught by TypeScript" are false.

Reply:
ok i See.

Reply:
master was good, not sure how you ended up with that code.

https://github.com/cordova-rtc/cordova-plugin-iosrtc/blob/master/js/RTCPeerConnection.js#L366

```
if (self.remoteDescription && data.remoteDescription) {
				self.remoteDescription.type = data.remoteDescription.type;
				self.remoteDescription.sdp = data.remoteDescription.sdp;
			} else if (data.remoteDescription) {
				self.remoteDescription = new RTCSessionDescription(data.remoteDescription);
			}
```

Reply:
got it, fixed on master